skbio/tree/_nj.py

```diff
@@ -286,3 +287,332 @@ def _pair_members_to_new_node(dm, i, j, disallow_negative_branch_length):
         j_to_u = 0
 
     return i_to_u, j_to_u
+
+
+def nni(tree, dm, inplace=True):
+    r"""Perform nearest neighbor interchanges on a phylogenetic tree.
+
+    Parameters
+    ----------
+    tree : skbio.TreeNode
+        Input phylogenetic tree to be rearranged.
+    dm : skbio.DistanceMatrix
+        Input distance matrix containing distances between taxa.
+    inplace : bool, optional
+        Whether manipulate the tree in place (``True``, default) or return a
+        copy of the tree (``False``).
+
+    Returns
+    -------
+    TreeNode
+        Rearranged phylogenetic tree (if ``inplace`` is ``True``).
+
+    Notes
+    -----
+    NNI algorithm for minimum evolution problem on phylogenetic trees. It rearranges
+    an initial tree topology by performing subtree exchanges such that the distance
+    is minimized. This implementation is based on the FastNNI algorithm [1]_.
+
+    The input tree is required to be binary and rooted at a leaf node such that
+    there is a unique descendant from the root.
+
+    References
+    ----------
+    .. [1] Desper R, Gascuel O. Fast and accurate phylogeny reconstruction
+       algorithms based on the minimum-evolution principle. J Comput Biol.
+       2002;9(5):687-705. doi: 10.1089/106652702761034136. PMID: 12487758.
+
+    """
+    # Initialize and populate the average distance matrix
+    if not inplace:
+        tree = tree.copy()
+    if len(tree.root().children) != 1:
+        print("Could not perform NNI. Tree needs to be rooted at a leaf node")
+        return
+    for node in tree.non_tips():
+        if len(node.children) != 2:
+            print("Could not perform NNI. Tree needs to be a binary tree")
+            return
+    adm = _average_distance_matrix(tree, dm)
+    while True:
+        # create heap of possible swaps and then swapping subtrees
+        # until no more swaps are possible.
+        adm = _average_distance_matrix(tree, dm)
```

Comment:
I noticed that both `_average_distance_matrix` and `_swap_heap` involve counting tips in the tree (`tree.root().count(tips=True)`). Counting involves a postorder traversal of the entire tree, which is expensive. Since the tree size doesn't change during the swapping process (if I understand correctly), it may be more efficient to pre-calcuate tip count just once and pass it to the helper function.